# Merging CMU dataset with the IMDb dataset

## Motivation

Why do we want to do that? To get clean metadata.

<div class='alert-warning'> Motivate the merge. </div>

## Methods

### Naive approach

To merge the CMU dataset with the IMDb dataset, we have to have a combination of common columns that uniquely identifies a movie on both datasets. 

The easiest is to merge them on the movie title.

<div class='alert-warning'> Here we show why merging on the title is problematic. </div>

The combination of the movie title and the year can be unique.

<div class='alert-warning'> Here we show why this is also problematic. </div>

### Crawling Wikipedia and querying Wikidata

The movies in the CMU dataset are linked with the Wikipedia ID and the Freebase ID. In the IMDb dataset, they are identified by the IMDb IDs. If only there existed a mapping between Wikipedia ID / Freebase ID to IMDb IDs, the merging could be done. First, we check that these IDs can uniquely identify a movie:

<div class='alert-warning'> ... </div>

We notice that the same string also appears in the URL of the movie page on the IMDb website. With the help of an external library, [wikipedia](https://wikipedia.readthedocs.org/en/latest/), we can get the content of the Wikipedia page of a movie from its Wikipedia ID in the CMU dataset. We also notice that the IMDb page of the movie is most of the times referenced in the Wikipedia pages of the movies, meaning that we can link the two by crawling Wikipedia. However, this approach might fail if the IMDb page is not included anywhere in the Wikipedia page, or the Wikipedia page is not retrievable from its page ID.

For these cases, we follow an alternative approach. In the Wikidata page of a movie, both the Freebase ID and the IMDb ID are listed. We can use the [Wikidata Query Service](https://query.wikidata.org) to match these two together.

The `helpers.crawl_wikipedia` and `helpers.crawl_wikidata` methods are used in `helpers.extract_cmu_imdb_mapping` to get such mappings. The whole process takes around 24 hours but can be improved by engineering the requests sent to Wikipedia by the external library. Since we only need to do this once, we opt to focus on the other parts and stick with the implementation of the external library. `helpers.extract_cmu_imdb_mapping` generates such mappings by the two methods, aggregates them, and stores the final mapping (and the most complete one) in `./generated/wp2imdb.csv`. The file is available in the repository but can be regenerated simply by running this function.

We use this mapping in the rest of this section to merge the two datasets.

In [1]:
import requests
import pandas as pd
from time import sleep
import wikipedia

In [15]:
pd.set_option("display.max_rows", None)
pd.set_option("display.max_colwidth", None)

## Analyze the generated mappings

In [2]:
mapping_01 = pd.read_csv('generated/wp2imdb_01.csv')
mapping_02 = pd.read_csv('generated/wp2imdb_02.csv')
mapping = pd.read_csv('generated/wp2imdb.csv')
cmu_movies = pd.read_csv('data/MovieSummaries/movie.metadata.tsv', sep='\t', usecols=[0, 1, 2], names=['wikipedia', 'freebase', 'title'])

In [3]:
len(mapping_01), len(mapping_02), len(mapping.imdb.unique()), len(cmu_movies)

(72180, 73894, 76885, 81741)

There are 57 IMDb movies that are duplicated in the first method, and 11 in the second one:

In [12]:
duplicates_01 = mapping_01[mapping_01.imdb.duplicated(keep=False)]
duplicates_02 = mapping_02[mapping_02.imdb.duplicated(keep=False)]

len(duplicates_01.imdb.unique()), len(duplicates_02.imdb.unique())

(57, 11)

Let's analyze the two mappings separately. First, we merge the duplicated mappings with the CMU table to get the titles. Then we add the URL of the Wikipedia page to check them more easily.

In [17]:
duplicates_01 = pd.merge(left=duplicates_01, right=cmu_movies.drop('freebase', axis=1), on='wikipedia', how='left').sort_values(by='imdb')
duplicates_01['url'] = duplicates_01.wikipedia.apply(lambda pageid: wikipedia.page(pageid=pageid).url)
display(duplicates_01, display_id=False)

wikipedia       imdb  \
26    29912713  tt0011325   
101    1364238  tt0011325   
6      7971186  tt0021644   
90     7531222  tt0021644   
34    26192132  tt0025472   
77     1416847  tt0025472   
53    35030671  tt0026191   
28    14711494  tt0026191   
92    22610953  tt0030337   
72    11635934  tt0030337   
24     2194415  tt0031077   
0     28322216  tt0031077   
89     4598097  tt0032264   
33     4598013  tt0032264   
69      264869  tt0033623   
94    23986728  tt0033623   
106   21954098  tt0034263   
31    31913677  tt0034263   
91    24732676  tt0034716   
66    10203059  tt0034716   
87    18271027  tt0039484   
108   18013931  tt0039484   
59    29141300  tt0040304   
38    29142474  tt0040304   
82    29166527  tt0042193   
76    29165415  tt0042193   
42    29245829  tt0044914   
56    29248702  tt0044914   
25    29250476  tt0045293   
37    29250219  tt0045293   
79    17980259  tt0046441   
99    26461347  tt0046441   
3     28089857  tt0050552   
54    28090138  tt0050552   
35    25665895  tt0051964   
102    1245942  tt0051964   
52     3371654  tt0052330   
83     1774116  tt0052330   
57    30290209  tt0055092   
68    26364509  tt0055092   
110   14748669  tt0056171   
22     1952819  tt0056171   
46    13550013  tt0059175   
15    22693744  tt0059175   
103   23666695  tt0059741   
51      145640  tt0059741   
55       30275  tt0063350   
43    32956666  tt0063350   
60    27486017  tt0064328   
85     2338179  tt0064328   
40    13098785  tt0065054   
30     1721414  tt0065054   
50     3605117  tt0065416   
32      183741  tt0065416   
93     5501808  tt0083015   
23      492200  tt0083015   
8      5297292  tt0084597   
19    35346685  tt0084597   
12       31629  tt0085630   
95    17565442  tt0085630   
96    16106115  tt0088445   
70    23293031  tt0088445   
62      309070  tt0089026   
104   23913058  tt0089026   
36     3624149  tt0092842   
48     4327278  tt0092842   
111   23315698  tt0092965   
98      146077  tt0092965   
65    23805173  tt0095344   
16     4237665  tt0095344   
29      213270  tt0096874   
71     1640125  tt0096874   
27     1380383  tt0104536   
1     17864265  tt0104536   
86    24936868  tt0113948   
18     3066864  tt0113948   
67      406339  tt0115986   
61      877220  tt0115986   
2       414113  tt0120737   
113     173941  tt0120737   
107   33215383  tt0138360   
7     35248640  tt0138360   
41     9912487  tt0183587   
21     9915882  tt0183587   
63     5040040  tt0206341   
4      1245381  tt0206341   
78    14983036  tt0218277   
80    34600272  tt0218277   
81    31767008  tt0250469   
17    15161056  tt0250469   
49    32523821  tt0358914   
44    25504132  tt0358914   
112   22611251  tt0388760   
97     4471177  tt0388760   
14     4218807  tt0944834   
74    19843601  tt0944834   
11    22132319  tt1056442   
105    9817916  tt1056442   
39     2511602  tt1080926   
5     23200481  tt1080926   
75     8827343  tt1167660   
45    24879053  tt1167660   
64    20030612  tt1247640   
73     3318002  tt1247640   
9     32631472  tt1258166   
10     2252617  tt1258166   
13     2560030  tt1453409   
109   31787087  tt1453409   
20    32802295  tt1535550   
84    13734892  tt1535550   
58    36281191  tt1954470   
47    31439778  tt1954470   
88    27019957  tt2382698   
100   31156288  tt2382698   

                                                       title  \
26                                            If I Were King   
101                                           If I Were King   
6                                             Laughing Gravy   
90                                                   Be Big!   
34                                             Marie Galante   
77                                   The Power and the Glory   
53                                          The Flying Fleet   
28                                              Ceiling Zero   
92                                            A Woman's Face   
72                           

[tt0104536](https://www.imdb.com/title/tt0104536/): Wikipedia page ([17864265](https://en.wikipedia.org/wiki/Itsy_Bitsy_Spider_(film))) corresponds to the right movie *Itsy Bitsy Spider*. Wikipedia page ([1380383](https://en.wikipedia.org/wiki/Bebe%27s_Kids)) corresponds to a movie called *Babe's Kids*. The reason for finding the IMDb ID is that in this Wikipedia page, there is an external link to *Itsy Bitsy Spider* in the References section, which has made this confusion.

We realize two things from this case:
1. There could be such mistakes in the first method for matching Wikipedia ID with IMDb ID.
2. It is more safe to stick with the second method.

> Most of the cases are like this.

> Easy fix is to loop over the external links in reverse order.

Now we do the same thing for the mapping from method 2 and we get the Wikidata pages:

In [31]:
Q = """
SELECT ?item WHERE {
  ?item wdt:P646 '%s' .
}
"""
URL = 'https://query.wikidata.org/sparql'

def get_wikidata_items(freebaseid):
    items = [
        b['item']['value']
        for b in requests.get(URL, params = {'format': 'json', 'query': Q % freebaseid}).json()['results']['bindings']
    ]
    sleep(5)
    return items

In [32]:
duplicates_02 = pd.merge(left=duplicates_02, right=cmu_movies.drop('wikipedia', axis=1), on='freebase', how='left').sort_values(by='imdb')
duplicates_02['url'] = duplicates_02.freebase.apply(get_wikidata_items)
display(duplicates_02, display_id=False)

freebase       imdb                                 title_x  \
0   /m/0gwygmm  tt0003886                             Enoch Arden   
1   /m/04csqxh  tt0003886                             Enoch Arden   
2   /m/05zkmzh  tt0004047                              Half Breed   
3   /m/05zm_7t  tt0004047                   The Conflicts of Life   
4   /m/06zm9kt  tt0044592                   Era lei che lo voleva   
5   /m/06zp6s1  tt0044592                              Oggi sposi   
6   /m/0283_5p  tt0080422                               Toothache   
7   /m/0283_05  tt0080422                          Dental Hygiene   
8    /m/0gxvwy  tt0102359                                Surprise   
9    /m/0gxvw7  tt0102359                           Light & Heavy   
11  /m/02qk_9g  tt0248428                                    Siva   
10   /m/02gvt2  tt0248428                                   Shiva   
12  /m/0h68hf7  tt0268117                           Anyay Abichar   
13  /m/0gx1gwf  tt0268117                                Aar Paar   
14  /m/0gg5yw0  tt0319813                             Ek Saudagar   
15  /m/02qkzwc  tt0319813                                  Mannan   
16  /m/0h1c674  tt0358914                     Maa avida Collector   
17  /m/09rxk5h  tt0358914                       Aavida Maa Aavide   
18   /m/0d7_8l  tt0815890                              Khan Kluay   
19  /m/051z4n5  tt0815890                                   Jumbo   
20  /m/063_93d  tt1283913         High School Musical: El Desafío   
21  /m/05znc1x  tt1283913  High School Musical: El Desafio Mexico   

                                   title_y  \
0                              Enoch Arden   
1                              Enoch Arden   
2                               Half Breed   
3                    The Conflicts of Life   
4                    Era lei che lo voleva   
5                               Oggi sposi   
6                                Toothache   
7                           Dental Hygiene   
8                                 Surprise   
9                            Light & Heavy   
11                                    Siva   
10                                   Shiva   
12                           Anyay Abichar   
13                                Aar Paar   
14                             Ek Saudagar   
15                                  Mannan   
16                     Maa avida Collector   
17                       Aavida Maa Aavide   
18                              Khan Kluay   
19                                   Jumbo   
20         High School Musical: El Desafío   
21  High School Musical: El Desafio Mexico   

                                     title  \
0                              Enoch Arden   
1                              Enoch Arden   
2                               Half Breed   
3                    The Conflicts of Life   
4                    Era lei che lo voleva   
5                               Oggi sposi   
6                                Toothache   
7                           Dental Hygiene   
8                                 Surprise   
9                            Light & Heavy   
11                                    Siva   
10                                   Shiva   
12                           Anyay Abichar   
13                                Aar Paar   
14                             Ek Saudagar   
15                                  Mannan   
16                     Maa avida Collector   
17                       Aavida Maa Aavide   
18                              Khan Kluay   
19                                   Jumbo   
20         High School Musical: El Desafío   
21  High School Musical: El Desafio Mexico   

                                           url  
0    [http://www.wikidata.org/entity/Q5379267]  
1     [http://www.wikidata.org/entity/Q629580]  
2    [http://www.wikidata.org/entity/Q4979898]  
3    [http://www.wikidata.org/entity/Q7435149]  
4    [http://www.wikidata.org/entity/Q3730996]  
5    [

By checking the Wikidata pages of these items, we realize that all these movies are closely related and are actually the same movie in IMDb most of the time, with different Freebase IDs. We conclude that this method is reliable.

## Merge

<div class='alert-warning'> Let's use method 2 for now. We will get 3k more movies but we need to re-run the first method (15 hours). </div>

In [49]:
mapping_02 = pd.read_csv('generated/wp2imdb_02.csv')
movies = pd.merge(left=cmu_movies, right=mapping_02, on='freebase', how='left')

In [50]:
movies.sample(10)

wikipedia    freebase                            title       imdb
41649    3838436   /m/0b2mcl                Follow Me Quietly  tt0041378
4344     2685725   /m/07xxt4  Sinbad and the Eye of the Tiger  tt0076716
74168    2388874   /m/078j5m              Jumborg Ace & Giant  tt3263160
37244   17061845  /m/041737v            Business and Pleasure  tt0021702
12603   35436294  /m/0j9mryz                       Jail Yatra  tt0390139
30437   27850610  /m/0cc5fmz                      Make a Wish        NaN
52172    5127062   /m/0d3_lf          Once Upon a Texas Train  tt0095781
69689   24485822  /m/080k97p                      Skiptracers  tt1270847
34724    1160944   /m/04cd1m                    Buddy the Gob  tt0024924
47034   18916948  /m/04jkrxj                  The Last Target  tt0068834